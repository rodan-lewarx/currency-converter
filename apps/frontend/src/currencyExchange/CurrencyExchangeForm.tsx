import { Currency } from "@task/currency-list-schema";
import { useReducer } from "react";
import { exchangeCurrency, amountSchema } from "./exchangeCurrency";
import { createDefaultState, formReducer } from "./usecurrencyExchangeForm";
import {
  Button,
  Error,
  Result,
  Input,
  Select,
  FormLayout,
  FormLabel,
} from "../ui-components";

export const CurrencyExchangeForm = ({
  currencyList,
}: {
  currencyList: Currency[];
}) => {
  const [state, dispatch] = useReducer(
    formReducer,
    createDefaultState(currencyList[0]),
  );

  const handleChangeAmount = (e: React.ChangeEvent<HTMLInputElement>): void =>
    dispatch({ type: "SET_AMOUNT", amount: e.target.value });

  const handleChangeCurrency = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const selectedCurrency = currencyList.find(
      (currency) => currency.currencyCode === e.target.value,
    )!;
    dispatch({ type: "SET_CURRENCY", currency: selectedCurrency });
  };

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const validatedAmount = amountSchema.safeParse(state.amount);
    if (!validatedAmount.success) {
      dispatch({
        type: "VALIDATE_FORM",
        error: validatedAmount.error.issues[0]?.message,
        result: "",
      });
    } else {
      const result = exchangeCurrency(validatedAmount.data, state.currency);
      dispatch({
        type: "VALIDATE_FORM",
        error: "",
        result: result,
      });
    }
  };

  return (
    <>
      <form onSubmit={handleSubmit}>
        <FormLayout>
          <Input
            type="number"
            placeholder="Enter amount"
            value={state.amount}
            onChange={handleChangeAmount}
          />
          <FormLabel>CZK to</FormLabel>
          <Select
            value={state.currency.currencyCode}
            onChange={handleChangeCurrency}
          >
            {currencyList.map((currency) => (
              <option key={currency.currencyCode} value={currency.currencyCode}>
                {currency.currencyCode}
              </option>
            ))}
          </Select>
          <Button type="submit">Convert</Button>
        </FormLayout>
        {state.error && <Error>{state.error}</Error>}
        {state.result && (
          <Result>{`${state.amount} CZK = ${state.result} ${state.currency.currencyCode}`}</Result>
        )}
      </form>
    </>
  );
};
