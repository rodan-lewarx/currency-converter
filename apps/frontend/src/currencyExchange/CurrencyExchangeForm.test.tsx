import { describe, expect, it, afterEach } from "vitest";
import { CurrencyExchangeForm } from "./CurrencyExchangeForm";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/vitest";

/**
 * @vitest-environment jsdom
 */

const currencyList = [
  {
    country: "United Kingdom",
    currency: "pound sterling",
    amount: 1,
    currencyCode: "GBP",
    rate: 28.222,
  },
  {
    country: "United States",
    currency: "US Dollar",
    amount: 1,
    currencyCode: "USD",
    rate: 22.781,
  },
];

afterEach(async () => {
  cleanup(); // clear testing data after each test run
});

describe("CurrencyExchangeForm", () => {
  it("should render a form", () => {
    const form = render(<CurrencyExchangeForm currencyList={currencyList} />);
    expect(form).toMatchSnapshot();
  });

  it("should calculate exchange rate when filling in data", () => {
    render(<CurrencyExchangeForm currencyList={currencyList} />);
    const input = screen.getByPlaceholderText("Enter amount");
    fireEvent.change(input, { target: { value: "10" } });
    const select = screen.getByRole("combobox");
    fireEvent.change(select, { target: { value: "USD" } });
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(screen.getByText("10 CZK = 0.44 USD")).toBeInTheDocument();
  });

  it("should return error when data is invalid", () => {
    render(<CurrencyExchangeForm currencyList={currencyList} />);
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(
      screen.getByText("Amount must be positive number."),
    ).toBeInTheDocument();
  });
});
