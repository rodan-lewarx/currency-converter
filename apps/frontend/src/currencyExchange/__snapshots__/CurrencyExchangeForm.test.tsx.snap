// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CurrencyExchangeForm > should render a form 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="sc-fujznN cRhTLx"
        >
          <input
            class="sc-gKAaef gvGmMp"
            placeholder="Enter amount"
            type="number"
            value=""
          />
          <span
            class="sc-pNWRh lnoOud"
          >
            CZK to
          </span>
          <select
            class="sc-iCoHzw hvmJbe"
          >
            <option
              value="GBP"
            >
              GBP
            </option>
            <option
              value="USD"
            >
              USD
            </option>
          </select>
          <button
            class="sc-gtsqUy fphciq"
            type="submit"
          >
            Convert
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="sc-fujznN cRhTLx"
      >
        <input
          class="sc-gKAaef gvGmMp"
          placeholder="Enter amount"
          type="number"
          value=""
        />
        <span
          class="sc-pNWRh lnoOud"
        >
          CZK to
        </span>
        <select
          class="sc-iCoHzw hvmJbe"
        >
          <option
            value="GBP"
          >
            GBP
          </option>
          <option
            value="USD"
          >
            USD
          </option>
        </select>
        <button
          class="sc-gtsqUy fphciq"
          type="submit"
        >
          Convert
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
